sql/src/main/java/io/crate/planner/v2/UpdateConsumer.java

```diff
@@ -111,47 +113,84 @@ public PlannedAnalyzedRelation visitUpdateAnalyzedStatement(UpdateAnalyzedStatem
-                            .output(Lists.newArrayList((Symbol) uidReference));
-
-                    Map<String, Symbol> assignments = new HashMap<>(nestedAnalysis.assignments().size());
-                    for(Map.Entry<Reference, Symbol> entry : nestedAnalysis.assignments().entrySet()) {
-                        assignments.put(entry.getKey().info().ident().columnIdent().fqn(), entry.getValue());
-                    }
-
-                    UpdateProjection updateProjection = new UpdateProjection(
-                            contextBuilder.outputs().get(0),
-                            assignments,
-                            whereClauseContext.whereClause().version().orNull());
-
-                    CollectNode collectNode = PlanNodeBuilder.collect(
-                            tableInfo,
-                            whereClause,
-                            contextBuilder.toCollect(),
-                            ImmutableList.<Projection>of(updateProjection)
-                    );
-                    MergeNode mergeNode = PlanNodeBuilder.localMerge(
-                            ImmutableList.<Projection>of(localMergeProjection), collectNode);
-                    childNodes.add(ImmutableList.<DQLPlanNode>of(collectNode, mergeNode));
+                if (whereClauseContext.ids().size() >= 1
+                        && whereClauseContext.routingValues().size() == whereClauseContext.ids().size()) {
+                    childNodes.add(updateById(nestedAnalysis, tableInfo, whereClause, whereClauseContext));
+                } else {
+                    childNodes.add(updateByQuery(nestedAnalysis, tableInfo, whereClause, whereClauseContext));
                 }
             }
 
             return new UpdateNode(childNodes);
         }
 
+        private List<DQLPlanNode> updateByQuery(UpdateAnalyzedStatement.NestedAnalyzedStatement nestedAnalysis,
+                                               TableInfo tableInfo,
+                                               WhereClause whereClause,
+                                               WhereClauseContext whereClauseContext) {
+
+            if(whereClauseContext.whereClause().version().isPresent()){
+                VersionRewriter versionRewriter = new VersionRewriter();
+                Symbol whereClauseQuery = versionRewriter.rewrite(whereClause.query());
+                whereClause = new WhereClause(whereClauseQuery);
+            }
+
+            if (!whereClause.noMatch() || !(tableInfo.isPartitioned() && whereClause.partitions().isEmpty())) {
+                // for updates, we always need to collect the `_uid`
+                Reference uidReference = new Reference(
+                        new ReferenceInfo(
+                                new ReferenceIdent(tableInfo.ident(), "_uid"),
+                                RowGranularity.DOC, DataTypes.STRING));
+
+                PlannerContextBuilder contextBuilder = new PlannerContextBuilder()
```

Comment:
does it make sense to use the PlannerContextBuilder here? There is no `InputColumn` mangling here so it should be possible to use the uidReference as is, or am I missing something?